prow/hook/server_test.go

```diff
@@ -334,3 +386,180 @@ func TestNeedDemux(t *testing.T) {
 		})
 	}
 }
+
+type roundTripFunc func(req *http.Request) *http.Response
+
+// RoundTrip .
+func (f roundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req), nil
+}
+
+// newTestClient returns *http.Client with Transport replaced to avoid making real calls
+func newTestClient(fn roundTripFunc) *http.Client {
+	return &http.Client{
+		Transport: fn,
+	}
+}
+
+func TestDemuxEvent(t *testing.T) {
+
+	getSecret := func() []byte {
+		var repoLevelSecret = `
+'*':
+  - value: abc
+    created_at: 2019-10-02T15:00:00Z
+  - value: key2
+    created_at: 2020-10-02T15:00:00Z
+foo/bar:
+  - value: 123abc
+    created_at: 2019-10-02T15:00:00Z
+  - value: key6
+    created_at: 2020-10-02T15:00:00Z
+`
+		return []byte(repoLevelSecret)
+	}
+
+	externalPlugins := map[string][]plugins.ExternalPlugin{
+		"kubernetes/test-infra": {
+			{
+				Name:     "coffee",
+				Endpoint: "/coffee",
+			},
+		},
+		"kubernetes/kubernetes": {
+			{
+				Name:     "gumbo",
+				Endpoint: "/gumbo",
+				Events:   []string{"issue_comment"},
+			},
+		},
+		"kubernetes": {
+			{
+				Name:     "chicken",
+				Endpoint: "/chicken",
+				Events:   []string{"repository"},
+			},
+			{
+				Name:     "water",
+				Endpoint: "/water",
+			},
+			{
+				Name:     "chocolate",
+				Endpoint: "/chocolate",
+				Events:   []string{"pull_request", "issue_comment", "repository"},
+			},
+		},
+	}
+
+	// This is the SHA1 signature for payload "$BODY" and signature "abc"
+	// echo -n $BODY | openssl dgst -sha1 -hmac abc
+	const hmac string = "sha1=d5f926df2d39006bdb5b6acb18f8fcdebad7a052"
+	const body string = `{
+  "action": "edited",
+  "changes": {
+    "default_branch": {
+      "from": "master"
+    }
+  },
+  "repository": {
+    "full_name": "kubernetes/test-infra",
+    "default_branch": "master"
+  }
+}`
+
+	metrics := githubeventserver.NewMetrics()
+	pa := &plugins.ConfigAgent{}
+	pa.Set(&plugins.Configuration{
+		ExternalPlugins: externalPlugins,
+	})
+
+	var testcases = []struct {
+		name string
+
+		Method string
+		Header map[string]string
+		Body   string
+
+		ExpectedDispatch []string
+	}{
+		{
+			name: "Repository Event",
```

Comment:
So this is a made-up event type? Can we maybe rename the test and event type to clarify that and avoid breaking the testcase should github actually add a `repository` event type, something like "unknown event type gets dispatched to external plugin"?

Reply:
It's an actual event type, in fact, the very reason for this PR, as I mentioned in the description :)

https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#repository

Reply:
@alvaroaleman Do you want me to add another test case with the made-up event as well? Please let me know what you prefer and I'm happy to squash all of the commits.

Reply:
Added in 9454c43a44.